aws/lambda/opensearch-gha-jobs/test_lambda_function.py

```diff
@@ -0,0 +1,311 @@
+import json
+from unittest import TestCase
+from unittest.mock import Mock
+
+from lambda_function import (
+    extract_dynamodb_key,
+    extract_dynamodb_table,
+    remove_document,
+    to_number,
+    unmarshal,
+    upsert_document,
+)
+
+
+def test_extract_dynamodb_table():
+    cases = [
+        {
+            "arn": "",
+            "expected": None,
+            "description": "Invalid input - empty input",
+        },
+        {
+            "arn": "FOOBAR",
+            "expected": None,
+            "description": "Invalid input - not in ARN format",
+        },
+        {
+            "arn": "arn:aws:dynamodb:us-east-1:12345:table/torchci-workflow-job/stream/2022-01-14T01:31:51.775",
+            "expected": "torchci-workflow-job",
+            "description": "An event coming from DynamoDB",
+        },
+    ]
+
+    for case in cases:
+        arn = case["arn"]
+        expected = case["expected"]
+        assert expected == extract_dynamodb_table({"eventSourceARN": arn})
```

Comment:
why are you using the raw python assert instead of unittest's `self.assertEquals`?

Sample code:
https://github.com/pytorch/pytorch/blob/main/test/test_content_store.py#L42

Reply:
Ah I can use `assertEquals`, but as I don't need other more advance asserts, thus I think it would be easier to just stick with the default `assert`.  Let me switch to `assertEquals`

Reply:
TIL, pytest doc https://docs.pytest.org/en/latest/getting-started.html#our-first-test-run mentions this.  Now I understand why all pytest examples simply use `assert`.  I think we could do the same here as I'm not really using `unittest` (besides mocking)

> You can use the assert statement to verify test expectations. pytest’s [Advanced assertion introspection](https://docs.python.org/3/reference/simple_stmts.html#assert) will intelligently report intermediate values of the assert expression so you can avoid the many names [of JUnit legacy methods](https://docs.python.org/3/library/unittest.html#testcase-objects).

Reply:
I poked around to see what this introspection tool's output looks like. It doesn't seem to quite work?

```
import unittest

class TestParsePrevTests(unittest.TestCase):

    def test_foo(self):
        a = "good"
        b = "bad"
        assert a == b, "oops"


if __name__ == "__main__":
    unittest.main()
```

Resulted in:

```
(base) ➜  pytorch git:(zainr/td-framework) ✗ python tools/test/test_heuristics.py
F
======================================================================
FAIL: test_foo (__main__.TestParsePrevTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/zainr/pytorch/tools/test/test_heuristics.py", line 8, in test_foo
    assert a == b, "oops"
AssertionError: oops

```

Now, if I did something wrong and it's actually supposed to work like [this StackOverflow question ](https://stackoverflow.com/questions/41522767/pytest-assert-introspection-in-helper-function) implies, it still isn't as nice and clean of an error message as the normal `assertEquals` function provides.  And if the code ever gets refactored into a helper method, then the assertion's fanciness will stop working (unless more special handling is added)

I'd suggest optimizing for engineering ease instead of minimizing package import overhead.  Having a consistent pattern for asserts will make it easy for folks to always do the right thing and not accidentally choose the wrong tool at a particular moment (and then spend time debugging why something isn't working the way they expected it to)

Reply:
To get the benefit of pytest here, you would need to run the test with it.  Take the same example `TestParsePrevTests` above, here is how the output looks like with pytest run with `pytest debug.py` or `python -mpytest debug.py`:

```
=================================================================== test session starts ====================================================================
platform darwin -- Python 3.9.17, pytest-7.2.0, pluggy-1.0.0
rootdir: /Users/huydo/Storage/mine/pytorch, configfile: pytest.ini
plugins: xdoctest-1.1.0, cpp-2.3.0, resume-0.0.1, rerunfailures-10.3, shard-0.1.2, flakefinder-1.1.0, hypothesis-6.56.4, xdist-3.0.2, profiling-1.7.0, repeat-0.9.1, anyio-3.6.2
collected 1 item
Running 1 items in this shard

debug.py F                                                                                                                                           [100%]

========================================================================= FAILURES =========================================================================
_______________________________________________________________ TestParsePrevTests.test_foo ________________________________________________________________
Traceback (most recent call last):
  File "/Users/huydo/Storage/mine/pytorch/debug.py", line 8, in test_foo
    assert a == b, "oops"
AssertionError: oops
assert 'good' == 'bad'
  - bad
  + good
================================================================= short test summary info ==================================================================
FAILED debug.py::TestParsePrevTests::test_foo - AssertionError: oops
==================================================================== 1 failed in 0.23s =====================================================================
```

On the other hand, running it with python `python debug.py` only yields the simplified output, thus the differences.  Nevertheless, using `assertEqual` is fine if we prefers running the test with plain python like what we used to do.

```
F
======================================================================
FAIL: test_foo (__main__.TestParsePrevTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/huydo/Storage/mine/pytorch/debug.py", line 8, in test_foo
    assert a == b, "oops"
AssertionError: oops

----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (failures=1)
```

Reply:
thank you!